rfc/0003-tests-arch.md

```diff
@@ -0,0 +1,180 @@
+It should test individual pieces of code using the tools available for each
+programming language.
+
+When implemented the developer should run all the test with a single command like:
+
+```bash
+# code code code
+make unit-tests
+```
+
+These tests should be run on every change in the repository (e.g. push, pull request).
+
+## Component tests
+
+These are the tests that validates each Kubewarden component behaviour
+independently from the other stack components.
+
+To test the Kubewarden controller, it's possible to use the test infrastructure
+from Kubebuilder. It's setup an test enviroment to validates the controller
+behaviour using Ginkgo and Gomega. [1](https://github.com/Azure/azure-databricks-operator/blob/0f722a710fea06b86ecdccd9455336ca712bf775/controllers/suite_test.go)
+[2](https://github.com/Azure/azure-databricks-operator/blob/0f722a710fea06b86ecdccd9455336ca712bf775/controllers/secretscope_controller_test.go)
+[3](https://book.kubebuilder.io/cronjob-tutorial/writing-tests.html)
+
+For the policy server and other Rust components, we can use the language features
+to integrations tests.[3](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html)
+
+When implemented the developer should run all the test with a single command like:
+
+```bash
+# code code code
+make acceptance-tests
+# or
+make integration-tests
+```
+
+These tests should also run on every change proposed in the component
+repository (e.g. push, pull request).
+
+
+## End-to-end tests
+
+These are the most complex tests. They will simulate what Kubewarden users will
+do to install, upgrade, configure and possibly uninstall Kubewarden stack.
+
+In these tests, different components versions would be tested together.
+Therefore, ensuring that the user will not face issues when using different
+components configuration and versions together. This should ensure that
+Kubernetes operator will have the expected features working.
+
+These tests are trigger when we have new releases (or release candidate) of
+any of the Kubewarden stack.  It should follow some steps:
+
+1. Setup Kubernetes cluster
+  Kubewarden is expected to run in multiple Kubernetes versions and machine
+  architectures. So, the end-to-end tests should ensure that all the supported
+  versions and architectures are tested.
+2. Install Kuberwarden stack on the cluster from step 1
+  The version to be tested should be installed. This means that when a component
+  is released. It should be tested against all the other components versions which
+  it supposed to work together.
+3. Run basic end-to-end tests
+  Validate if the basic features are working as expected
+4. Run reconfiguration tests
+  Validates if the user is able to reconfigure a Kubewarden stack installed
+5. Run upgrade tests
+  Validate if the user is able to upgrade and continue to work with the latest release.
+
+Steps 3,4 and 5 may run in parallel.
```

Comment:
What does this mean? They can run in parallel in different environments, or acting in the same environment? I think it's best to start with a simple baseline, where reconfigurations do not happen along with upgrades.

Reply:
I mean that we can run in 3 different clusters. Each one runs different set of tests (steps 3,4,5)